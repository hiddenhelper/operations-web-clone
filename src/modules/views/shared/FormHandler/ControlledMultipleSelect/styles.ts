import { createStyles, makeStyles } from '@material-ui/core/styles';
import { toREM } from '../../../../../utils/generalUtils';

export const useStyles = makeStyles(theme =>
  createStyles({
    selectLabel: {
      fontSize: '0.85rem',
      maxHeight: '20px',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap',
      overflow: 'hidden',
    },
    iconRoot: {
      fontSize: '1.3rem',
    },
    formGroupRoot: {
      display: 'grid',
      gridTemplateColumns: '1fr 1fr 1fr',
      gridTemplateRows: 'repeat(7, minmax(0, 1fr))',
      gridGap: '12px 15px',
      gridAutoRows: '2em',
      gridAutoFlow: 'column',

      '@media (max-width:1380px)': {
        gridTemplateColumns: '1fr 1fr',
        gridTemplateRows: 'repeat(auto-fit, minmax(0, 1fr))',
        gridAutoFlow: 'dense',
      },
      '@media (max-width:1200px)': {
        gridTemplateColumns: '1fr',
        gridTemplateRows: 'repeat(auto-fit, minmax(0, 1fr))',
        gridAutoFlow: 'dense',
      },

      '& .MuiFormControlLabel-root': {
        width: `${toREM(278)}`,
        maxHeight: '32px',
        margin: 0,

        '& .MuiSvgIcon-root': {
          fontSize: '1.5rem',
          marginRight: '4px',
        },
        '& .MuiTypography-root': {
          fontSize: `${toREM(15)}`,
        },
      },
    },
    verticalRows: {
      display: 'flex',
      flexDirection: 'column',
    },
    formControlRoot: {
      width: '100%',
    },
  })
);
