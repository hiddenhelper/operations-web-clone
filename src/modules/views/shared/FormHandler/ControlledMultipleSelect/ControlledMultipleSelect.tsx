import React, { memo, useCallback } from 'react';
import FormControl from '@material-ui/core/FormControl';
import FormGroup from '@material-ui/core/FormGroup';
import FormControlLabel from '@material-ui/core/FormControlLabel';

import ControlledCheckbox from '../Checkbox';
import { useStyles } from './styles';
import { getConditionalDefaultValue } from '../../../../../utils/generalUtils';

export interface IControlledMultipleSelectProps {
  name: string;
  items: { id: string; name: string }[];
  value: { id: string; name: string }[];
  styleClasses?: any;
  verticalRows?: boolean;
  onChange: (data: any) => void;
  error?: boolean;
}

const ControlledMultipleSelect = ({ items, name, value, styleClasses = {}, verticalRows = false, onChange, error = false }: IControlledMultipleSelectProps) => {
  const styles = useStyles();
  const onChangeHandler = useCallback(
    event => {
      let optList = [...value];
      const item = items.find(v => v.id === event.target.value);
      if (!!value.find(v => v.id === item.id)) {
        optList = [...optList.filter(optItem => optItem.id !== item.id)];
      } else {
        optList = [...optList, item];
      }
      onChange(/* istanbul ignore next */ prevState => ({ ...prevState, [name]: optList }));
    },
    [name, value, items, onChange]
  );
  return (
    <div>
      <FormControl component="fieldset" classes={{ root: styleClasses.formControlRoot ?? styles.formControlRoot }}>
        <FormGroup
          className={getConditionalDefaultValue(verticalRows, styles.verticalRows, '')}
          classes={{ root: styleClasses.formGroupRoot ?? styles.formGroupRoot }}
        >
          {items.map((item, index) => (
            <FormControlLabel
              key={index}
              control={
                <ControlledCheckbox isChecked={!!value?.find(v => v.id === item.id)} name={name} value={item.id} onChange={onChangeHandler} error={error} />
              }
              label={item.name}
              classes={{ label: styles.selectLabel }}
            />
          ))}
        </FormGroup>
      </FormControl>
    </div>
  );
};

export default memo(ControlledMultipleSelect);
