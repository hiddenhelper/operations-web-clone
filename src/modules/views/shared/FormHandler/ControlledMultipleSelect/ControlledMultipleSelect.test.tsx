import React from 'react';
import { act, fireEvent, render, RenderResult } from '@testing-library/react';

import ControlledMultipleSelect, { IControlledMultipleSelectProps } from './ControlledMultipleSelect';

describe('Controlled Multiple Select', () => {
  let wrapper: RenderResult;
  let props: IControlledMultipleSelectProps = {
    items: [],
    name: '',
    value: [],
    onChange: jest.fn(),
  };

  it('should render', () => {
    wrapper = render(<ControlledMultipleSelect {...props} />);
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should handle item change', async () => {
    props = {
      items: [
        {
          id: '1',
          name: 'item1',
        },
        {
          id: '2',
          name: 'item2',
        },
      ],
      name: 'test',
      onChange: jest.fn(),
      value: [
        {
          id: '1',
          name: 'item1',
        },
      ],
    };
    wrapper = render(<ControlledMultipleSelect {...props} />);

    await act(async () => {
      fireEvent.click(wrapper.getAllByTestId('multiple-checkbox')[0].querySelector('input'));
    });

    expect(props.onChange).toHaveBeenCalled();
    await act(async () => {
      fireEvent.click(wrapper.getAllByTestId('multiple-checkbox')[1].querySelector('input'));
    });

    expect(props.onChange).toHaveBeenCalled();
  });
});
