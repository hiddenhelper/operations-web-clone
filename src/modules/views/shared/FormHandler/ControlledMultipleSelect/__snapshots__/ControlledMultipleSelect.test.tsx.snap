// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controlled Multiple Select should render 1`] = `
<div>
  <div>
    <fieldset
      class="MuiFormControl-root makeStyles-formControlRoot-5"
    >
      <div
        class="MuiFormGroup-root makeStyles-formGroupRoot-3"
      />
    </fieldset>
  </div>
</div>
`;
